util/pingpong.c

```diff
@@ -0,0 +1,1940 @@
+		if (ret)
+			FT_PRINTERR("fi_cntr_wait", ret);
+	} else {
+		FT_ERR("Trying to get a TX completion when no TX CQ or counter were opened");
+		ret = -FI_EOTHER;
+	}
+	return ret;
+}
+
+int ft_cq_readerr(struct fid_cq *cq)
+{
+	struct fi_cq_err_entry cq_err;
+	int ret;
+
+	ret = fi_cq_readerr(cq, &cq_err, 0);
+	if (ret < 0) {
+		FT_PRINTERR("fi_cq_readerr", ret);
+	} else {
+		FT_CQ_ERR(cq, cq_err, NULL, 0);
+		ret = -cq_err.err;
+	}
+	return ret;
+}
+
+
+int ft_sync(struct ct_pingpong *ct)
+{
+	int ret;
+
+	if (ct->opts.dst_addr) {
+		ret = ft_tx(ct, ct->ep, 1);
+		if (ret)
+			return ret;
+
+		ret = ft_rx(ct, ct->ep, 1);
+	} else {
+		ret = ft_rx(ct, ct->ep, 1);
+		if (ret)
+			return ret;
+
+		ret = ft_tx(ct, ct->ep, 1);
+	}
+
+	return ret;
+}
+
+#define FT_EP_BIND(ep, fd, flags)					\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_ep_bind((ep), &(fd)->fid, (flags));	\
+			if (ret) {					\
+				FT_PRINTERR("fi_ep_bind", ret);		\
+				return ret;				\
+			}						\
+		}							\
+	} while (0)
+
+int ft_init_alias_ep(struct ct_pingpong *ct, uint64_t flags)
+{
+	int ret;
+	ret = fi_ep_alias(ct->ep, &(ct->alias_ep), flags);
+	if (ret) {
+		FT_PRINTERR("fi_ep_alias", ret);
+		return ret;
+	}
+	return 0;
+}
+
+int ft_init_ep(struct ct_pingpong *ct)
+{
+	int flags, ret;
+
+	if (ct->fi->ep_attr->type == FI_EP_MSG)
+		FT_EP_BIND(ct->ep, ct->eq, 0);
+	FT_EP_BIND(ct->ep, ct->av, 0);
+	FT_EP_BIND(ct->ep, ct->txcq, FI_TRANSMIT);
+	FT_EP_BIND(ct->ep, ct->rxcq, FI_RECV);
+
+	/* TODO: use control structure to select counter bindings explicitly */
+	flags = !ct->txcq ? FI_SEND : 0;
+	if (ct->hints->caps & (FI_WRITE | FI_READ))
+		flags |= ct->hints->caps & (FI_WRITE | FI_READ);
+	else if (ct->hints->caps & FI_RMA)
+		flags |= FI_WRITE | FI_READ;
+	FT_EP_BIND(ct->ep, ct->txcntr, flags);
+	flags = !ct->rxcq ? FI_RECV : 0;
+	if (ct->hints->caps & (FI_REMOTE_WRITE | FI_REMOTE_READ))
+		flags |= ct->hints->caps & (FI_REMOTE_WRITE | FI_REMOTE_READ);
+	else if (ct->hints->caps & FI_RMA)
+		flags |= FI_REMOTE_WRITE | FI_REMOTE_READ;
+	FT_EP_BIND(ct->ep, ct->rxcntr, flags);
+
+	ret = fi_enable(ct->ep);
+	if (ret) {
+		FT_PRINTERR("fi_enable", ret);
+		return ret;
+	}
+
+	if (ct->fi->rx_attr->op_flags != FI_MULTI_RECV) {
+		/* Initial receive will get remote address for unconnected EPs */
+		ret = ft_post_rx(ct, ct->ep, MAX(ct->rx_size, FT_MAX_CTRL_MSG), &(ct->rx_ctx));
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+
+int ft_av_insert(struct fid_av *av, void *addr, size_t count, fi_addr_t *fi_addr,
+		uint64_t flags, void *context)
+{
+	int ret;
+
+	ret = fi_av_insert(av, addr, count, fi_addr, flags, context);
+	if (ret < 0) {
+		FT_PRINTERR("fi_av_insert", ret);
+		return ret;
+	} else if (ret != count) {
+		FT_ERR("fi_av_insert: number of addresses inserted = %d;"
+			       " number of addresses given = %zd\n", ret, count);
+		return -EXIT_FAILURE;
+	}
+
+	return 0;
+}
+
+/* TODO: retry send for unreliable endpoints */
```

Comment:
Yes, you should do that. 😄 

Maybe have a 2nd RDM endpoint for doing these kinds of control messages?